src/commando/impl/utils.cljc

```
@@ -19,25 +22,83 @@
+               :error-data-string false
+               :hook-execute-start (fn [e] (println (:uuid e)))
+               :hook-execute-end (fn [e] (println (:uuid e) (:stats e)))}]
+       (commando.core/execute
+         [commando.commands.builtin/command-from-spec]
+         {\"1\" 1
+          \"2\" {:commando/from [\"1\"]}
+          \"3\" {:commando/from [\"2\"]}}))"
   -execute-config-default)
 
+(def ^:dynamic
+  *execute-internals*
+  "Dynamic variable to keep context information about the execution
+   setup.
+   - `:uuid` the unique name of execution, generated everytime the user
+     invoke `commando.core/execute`
+   - `:stack` in case of user use `commando.commands.builtin/command-macro-spec`,
+     or `commando.commands.query-dsl/command-resolve-json-spec` or any sort of
+     commands what invoking `commando.core/execute` inside of parent instruction
+     by simulation recursive call, the :stack key will store the invocation stack
+     in vector of :uuids"
+  {:uuid nil
+   :stack []})
+
+(defn -execute-internals-push
+  "Update *execute-internals* structure"
+  [uuid-execute-identifier]
+  (-> *execute-internals*
+    (assoc :uuid uuid-execute-identifier)
+    (update :stack conj uuid-execute-identifier)))
+
 (defn execute-config
   "Returns the effective configuration for `commando/execute`, getting data from dynamic variable `commando.impl.utils/*execute-config*`"
   []
   (merge -execute-config-default *execute-config*))
 
-(def ^{:dynamic true
-       :private true
-       :doc "For debugging purposes and some mysterious reason of setting it dynamically during execution"}
+(defn hook-process
+  "Function will handle a hooks passed from users.
+   Available hooks:
+    - `:hook-execute-start`,
+    - `:hook-execute-end`.
+
+   Read more:
+      `commando.impl.utils/*execute-config*`"
+  [status-map hook]
+  (when hook
+    (try
+      (hook status-map)
+      (catch #?(:clj Exception
+                :cljs :default) e
+        nil)))
```

Hmmm, I'm not so sure that we should silently swallow this error. It will be surprising to use it and has no error and no information what happened.

---

Also as I'm thinking about it, what do you think about doing middleware pattern here instead of a hook? Just like in ring in the V pattern, middlewares called before execute and than after execute by wrapping it.